src/Symfony/Component/Validator/Constraints/Compound.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\Exception\ConstraintDefinitionException;
+
+/**
+ * Extend this class to create a reusable set of constraints.
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ */
+abstract class Compound extends Composite
+{
+    /** @var Constraint[] */
+    public $constraints = [];
+
+    public function __construct($options = null)
+    {
+        if (isset($options[$this->getCompositeOption()])) {
+            throw new ConstraintDefinitionException(sprintf('You can\'t redefine the "%s" option. Use the %s::getConstraints() method instead.', $this->getCompositeOption(), __CLASS__));
+        }
+
+        if (0 === \count($constraints = $this->getConstraints($this->normalizeOptions($options)))) {
+            throw new ConstraintDefinitionException(sprintf('The %s::getConstraints() method must return at least one constraint.', __CLASS__));
+        }
+
+        $this->constraints = $constraints;
+
+        parent::__construct($options);
+    }
+
+    final protected function getCompositeOption()
+    {
+        return 'constraints';
+    }
+
+    final public function validatedBy()
+    {
+        return CompoundValidator::class;
+    }
+
+    abstract protected function getConstraints(array $options): array;
```

Comment:
Instead of exposing the normalized options as a method argument, we could rely on populated instance properties by calling the `Constraint::__construct` beforehand (as in original commit https://github.com/symfony/symfony/pull/34334/commits/9fab172dfa471a7ae29afd5ef63f6f42bafdf2cb#diff-fa2f43b5bc196bf979f6e59ec699e462R33-R41). But this would call this construct twice as it'll be called in parent one too. This can either be workaround to ensure this constructor is called once, or solved in a totally different fashion (don't extend `Composite` but used a shared trait?)

Reply:
missing `@return Constraint[]`